src/MongoDB.Driver.Core/Core/Authentication/ScramSha256Authenticator.cs

```diff
@@ -0,0 +1,116 @@
+﻿/* Copyright 2015-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Security.Cryptography;
+using System.Text;
+using MongoDB.Bson.IO;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// A SCRAM-SHA256 SASL authenticator.
+    /// </summary>
+    public sealed class ScramSha256Authenticator : ScramShaAuthenticator
+    {
+        // static properties
+        /// <summary>
+        /// Gets the name of the mechanism.
+        /// </summary>
+        /// <value>
+        /// The name of the mechanism.
+        /// </value>
+        public static string MechanismName => "SCRAM-SHA-256";
+
+        // constructors
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ScramSha256Authenticator"/> class.
+        /// </summary>
+        /// <param name="credential">The credential.</param>
+        public ScramSha256Authenticator(UsernamePasswordCredential credential)
+            : this(credential, new DefaultRandomStringGenerator()) { }
+
+        internal ScramSha256Authenticator(UsernamePasswordCredential credential, IRandomStringGenerator randomStringGenerator)
+            : base(credential, HashAlgorithmName.SHA256, randomStringGenerator, H256, Hi256, Hmac256) { }
+
+        private static byte[] H256(byte[] data)
+        {
+            using (var sha256 = SHA256.Create())
+            {
+                return sha256.ComputeHash(data);
+            }
+        }
+
+        private static byte[] Hi256(UsernamePasswordCredential credential, byte[] salt, int iterations)
+        {
+#if NET45
+            var passwordIntPtr = Marshal.SecureStringToGlobalAllocUnicode(credential.Password);
+#else
+            var passwordIntPtr = SecureStringMarshal.SecureStringToGlobalAllocUnicode(credential.Password);
```

Comment:
As an aside, some articles suggest that SecureString is more effort than it's worth: perhaps a discussion worth having in the future.